File: packages/email/src/tokens.ts
export type UnsubscribeCategory = "marketing" | "transactional";

type TokenPayload = {
  userId: string;
  category: UnsubscribeCategory;
  iat: number;
};

function base64urlEncode(data: Uint8Array): string {
  const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  let result = "";
  const bytes = data;
  for (let i = 0; i < bytes.length; i += 3) {
    const b0 = bytes[i] ?? 0;
    const b1 = bytes[i + 1] ?? 0;
    const b2 = bytes[i + 2] ?? 0;
    result += chars[b0 >> 2];
    result += chars[((b0 & 3) << 4) | (b1 >> 4)];
    result += i + 1 < bytes.length ? chars[((b1 & 15) << 2) | (b2 >> 6)] : "=";
    result += i + 2 < bytes.length ? chars[b2 & 63] : "=";
  }
  return result.replace(/\+/g, "-").replace(/\//g, "_").replace(/=/g, "");
}

function base64urlDecode(str: string): Uint8Array {
  const base64 = str.replace(/-/g, "+").replace(/_/g, "/");
  const padded = base64.padEnd(base64.length + ((4 - (base64.length % 4)) % 4), "=");
  const binary = atob(padded);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

async function importKey(secret: string): Promise<CryptoKey> {
  const encoder = new TextEncoder();
  return crypto.subtle.importKey(
    "raw",
    encoder.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign", "verify"],
  );
}

export async function generateUnsubscribeToken(
  userId: string,
  category: UnsubscribeCategory,
  secret: string,
): Promise<string> {
  const payload: TokenPayload = { userId, category, iat: Math.floor(Date.now() / 1000) };
  const payloadJson = JSON.stringify(payload);
  const encoder = new TextEncoder();
  const payloadBytes = encoder.encode(payloadJson);
  const payloadB64 = base64urlEncode(payloadBytes);

  const key = await importKey(secret);
  const signatureBuffer = await crypto.subtle.sign("HMAC", key, encoder.encode(payloadB64));
  const signatureB64 = base64urlEncode(new Uint8Array(signatureBuffer));

  return `${payloadB64}.${signatureB64}`;
}

export async function verifyUnsubscribeToken(
  token: string,
  secret: string,
  maxAgeSeconds: number = 30 * 24 * 3600,
): Promise<{ userId: string; category: UnsubscribeCategory } | null> {
  const parts = token.split(".");
  if (parts.length !== 2) return null;

  const [payloadB64, signatureB64] = parts as [string, string];

  let payloadJson: string;
  let payload: unknown;
  try {
    const payloadBytes = base64urlDecode(payloadB64);
    payloadJson = new TextDecoder().decode(payloadBytes);
    payload = JSON.parse(payloadJson) as unknown;
  } catch {
    return null;
  }

  const key = await importKey(secret);
  const encoder = new TextEncoder();

  let signatureBytes: Uint8Array;
  try {
    signatureBytes = base64urlDecode(signatureB64);
  } catch {
    return null;
  }

  const valid = await crypto.subtle.verify(
    "HMAC",
    key,
    signatureBytes.buffer as ArrayBuffer,
    encoder.encode(payloadB64),
  );

  if (!valid) return null;

  if (
    typeof payload !== "object" ||
    payload === null ||
    !("userId" in payload) ||
    !("category" in payload) ||
    typeof (payload as Record<string, unknown>).userId !== "string" ||
    ((payload as Record<string, unknown>).category !== "marketing" &&
      (payload as Record<string, unknown>).category !== "transactional")
  ) {
    return null;
  }

  const typed = payload as TokenPayload;
  const nowSeconds = Math.floor(Date.now() / 1000);

  if (
    typeof typed.iat !== "number" ||
    !Number.isFinite(typed.iat) ||
    typed.iat > nowSeconds ||
    nowSeconds - typed.iat > maxAgeSeconds
  ) {
    return null;
  }

  return { userId: typed.userId, category: typed.category };
}

File: py/ventora_email/src/ventora_email/tokens.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import math
import time
from typing import Literal

UnsubscribeCategory = Literal["marketing", "transactional"]


def generate_unsubscribe_token(
    user_id: str,
    category: UnsubscribeCategory,
    secret: str,
) -> str:
    """Generate a stateless HMAC-SHA256 unsubscribe token."""
    payload = json.dumps({"userId": user_id, "category": category, "iat": int(time.time())})
    payload_b64 = base64.urlsafe_b64encode(payload.encode()).rstrip(b"=").decode()
    sig = hmac.new(secret.encode(), payload_b64.encode(), hashlib.sha256).digest()
    sig_b64 = base64.urlsafe_b64encode(sig).rstrip(b"=").decode()
    return f"{payload_b64}.{sig_b64}"


_DEFAULT_MAX_AGE_SECONDS = 30 * 24 * 3600  # 30 days


def verify_unsubscribe_token(
    token: str,
    secret: str,
    max_age_seconds: int = _DEFAULT_MAX_AGE_SECONDS,
) -> dict[str, str] | None:
    """Verify token signature and expiry, returning payload or None."""
    try:
        parts = token.split(".")
        if len(parts) != 2:
            return None
        payload_b64, sig_b64 = parts
        expected_sig = hmac.new(secret.encode(), payload_b64.encode(), hashlib.sha256).digest()
        expected_b64 = base64.urlsafe_b64encode(expected_sig).rstrip(b"=").decode()
        if not hmac.compare_digest(expected_b64, sig_b64):
            return None
        # Pad base64 if needed
        padding = 4 - len(payload_b64) % 4
        if padding != 4:
            payload_b64 += "=" * padding
        data = json.loads(base64.urlsafe_b64decode(payload_b64).decode())
        if not isinstance(data.get("userId"), str):
            return None
        if data.get("category") not in ("marketing", "transactional"):
            return None
        issued_at = data.get("iat")
        if (
            isinstance(issued_at, bool)
            or not isinstance(issued_at, int | float)
            or not math.isfinite(issued_at)
        ):
            return None
        now = time.time()
        if issued_at > now or now - issued_at > max_age_seconds:
            return None
        return {"userId": data["userId"], "category": data["category"]}
    except Exception:  # noqa: BLE001
        return None
